client/allocrunner/taskrunner/otel_hook.go

```diff
@@ -0,0 +1,144 @@
+package taskrunner
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	log "github.com/hashicorp/go-hclog"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"go.opentelemetry.io/otel/baggage"
+)
+
+const envKeyOtelResourceAttrs = "OTEL_RESOURCE_ATTRIBUTES"
+
+type otelHookConfig struct {
+	logger log.Logger
+	alloc  *structs.Allocation
+	node   *structs.Node
+}
+
+type otelHook struct {
+	alloc  *structs.Allocation
+	node   *structs.Node
+	logger log.Logger
+}
+
+func newOtelHook(config *otelHookConfig) *otelHook {
+	hook := &otelHook{
+		alloc: config.alloc,
+		node:  config.node,
+	}
+	hook.logger = config.logger.Named(hook.Name()).
+		With("alloc_id", config.alloc.ID)
+
+	return hook
+}
+
+func (h *otelHook) Name() string {
+	return "otel"
+}
+
+func (h *otelHook) Prestart(_ context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	logger := h.logger.With("task", req.Task.Name)
+
+	resourceAttrsEnv, ok := req.TaskEnv.EnvMap[envKeyOtelResourceAttrs]
+	if ok && resourceAttrsEnv == "" {
+		logger.Debug("skipping OTEL_RESOURCE_ATTRIBUTES environment variable")
+		return nil
+	}
+
+	resourceAttrs, err := generateBaggage(h.alloc, req.Task, h.node)
+	if err != nil {
+		logger.Warn("failed to generate OTEL_RESOURCE_ATTRIBUTES environment variable", "error", err)
+		return nil
+	}
+
+	if resourceAttrsEnv != "" {
+		logger.Debug("merging existing OTEL_RESOURCE_ATTRIBUTES environment variable values", "attrs", resourceAttrsEnv)
+
+		taskBaggage, err := baggage.Parse(resourceAttrsEnv)
+		if err != nil {
+			logger.Warn("failed to parse task environment variable OTEL_RESOURCE_ATTRIBUTES as baggage",
+				"otel_resource_attributes", resourceAttrsEnv, "error", err)
+		} else {
+			for _, m := range taskBaggage.Members() {
+				k, v := m.Key(), m.Value()
+				logger.Trace("found member", "key", k, "value", v)
+
+				// TODO(luiz): don't create new member once baggage.Members()
+				// returns values with `hasData` set to `true`.
+				// https://github.com/open-telemetry/opentelemetry-go/issues/3164
+				member, err := baggage.NewMember(k, v)
+				if err != nil {
+					logger.Warn("failed to create new baggage member", "key", k, "value", v, "error", err)
```

Comment:
The ridiculous name "baggage" aside, which isn't your fault, I'm not sure we should expose that inside-baseball terminology to end-users?

Reply:
Ah right, good point 👍